src/IB/IBFESurfaceMethod.cpp

```diff
@@ -640,48 +779,389 @@ IBFESurfaceMethod::interpolateVelocity(const int u_data_idx,
+                            {
+                                X_shifted[d] = x_qp[d + s * NDIM] + periodic_shifts[d + k * NDIM];
+                            }
+
+                            for (unsigned int d = 0; d < NDIM; ++d)
+                            {
+                                ic_center[d] = ilower[d] + NINT((X_shifted[d] - x_lower_axis[d]) / dx[d] - 0.5);
+                                X_cell[d] =
+                                    x_lower_axis[d] + (static_cast<double>(ic_center[d] - ilower[d]) + 0.5) * dx[d];
+
+                                if (X_shifted[d] <= X_cell[d])
+                                {
+                                    ic_lower[d] = ic_center[d] - 1;
+                                    ic_upper[d] = ic_center[d];
+                                }
+                                else
+                                {
+                                    ic_lower[d] = ic_center[d];
+                                    ic_upper[d] = ic_center[d] + 1;
+                                }
+                                ic_trimmed_lower[d] = std::max(ic_lower[d], ilower[d] - u_gcw[d]);
+                                ic_trimmed_upper[d] = std::min(ic_upper[d], iupper[d] + u_gcw[d]);
+                            }
+
+                            if (X_shifted[0] <= X_cell[0])
+                                w0[0] = (X_cell[0] - X_shifted[0]) / dx[0];
+                            else
+                                w0[0] = 1.0 + (X_cell[0] - X_shifted[0]) / dx[0];
+                            wr0[0] = w0[0];
+                            w0[1] = 1.0 - w0[0];
+                            wr0[1] = -w0[1];
+
+                            if (X_shifted[1] <= X_cell[1])
+                                w1[0] = (X_cell[1] - X_shifted[1]) / dx[1];
+                            else
+                                w1[0] = 1.0 + (X_cell[1] - X_shifted[1]) / dx[1];
+                            wr1[0] = w1[0];
+                            w1[1] = 1.0 - w1[0];
+                            wr1[1] = -w1[1];
+#if (NDIM == 3)
+                            if (X_shifted[2] <= X_cell[2])
+                                w2[0] = (X_cell[2] - X_shifted[2]) / dx[2];
+                            else
+                                w2[0] = 1.0 + (X_cell[2] - X_shifted[2]) / dx[2];
+                            wr2[0] = w2[0];
+                            w2[1] = 1.0 - w2[0];
+                            wr2[1] = -w2[1];
+#endif
+
+                            boost::multi_array<double, NDIM + 1> Ujump(
+                                boost::extents[range(ic_trimmed_lower[0], ic_trimmed_upper[0] + 1)]
+                                              [range(ic_trimmed_lower[1], ic_trimmed_upper[1] + 1)]
+#if (NDIM == 3)
+                                              [range(ic_trimmed_lower[2], ic_trimmed_upper[2] + 1)]
+#endif
+                                              [range(0, NDIM)]);
+
+#if (NDIM == 2)
```

Comment:
This needs some comments.

Also, could you avoid separate LL/LLL data structures by indexing on NodeIndex?

Reply:
I knew you're gonna say that:D ...okay I'll make it one index